File: wereyouhere/normalise.py
from re import compile as R

STRIP_RULES = [
    [R('.*')                     , R('^\\w+://'         )],
    [R('reddit.com|youtube.com') , R('(www|ww|amp)\\.'  )],
    [R('.*')                     , R('[&#].*$'       )],
    [
        [R('^youtube') , None],
        [R('.*')       , R('[\\?].*$')],
    ]
]


def normalise_url(url):
    cur = url
    for thing in STRIP_RULES:
        first = thing[0]
        rules = None
        if isinstance(first, list):
            rules = thing
        else:
            rules = [thing]

        for target, reg in rules:
            if target.search(cur):
                if reg is not None:
                    cur = reg.sub('', cur)
                break


    return cur


# use [] instead of () so it's easy to copy regexes to js
# TODO eh, None vs null..

# TODO fine tune, start with reddit?
# TODO ok, if first elemnent is a rule, apply it
# if only one, bail
# if it's a list, they are mutually exclusive?

File: extension/src/normalise.ts
import type {Url} from './common'

// TODO should probably be merged with common or something...

const
STRIP_RULES = [
    [[RegExp('.*')                     , RegExp('^\\w+://'         )]],
    [[RegExp('.*')                     , RegExp('(www|ww|amp)\\.'  )]],
    [[RegExp('.*')                     , RegExp('[&#].*$'          )]],
    [[RegExp('.*')                     , RegExp('/$'               )]],
]
// TODO perhaps that should be semi-configurable

// TODO maybe use that normalisation library and then adjust query params etc

/*
  I think, most common usecases are:
  - blacklisting whole domain (e.g. for privacy reasons, like bank/etc or if something is broken)
  - blacklisting specific pages (e.g. reddit/twitter/fb main page so it doesn't result it too many child contexts)
  For that current approach is fine.
*/

// TODO careful about dots etc?

export function normalise_url(url: string): string {
    let cur = url;
    STRIP_RULES.forEach((rules: Array<Array<RegExp | null>>) => { // meh impure foreach..
        for (const rule of rules) {
            const target: RegExp = rule[0]!
            const reg: RegExp | null = rule[1]
            if (cur.search(target) >= 0) {
                if (reg != null) {
                    cur = cur.replace(reg, '');
                }
                break;
            }
        }
    });
    return cur;
}

const _re = RegExp('^(www|ww|amp)\\.'  )
export function normaliseHostname(url: string): string {
    return url.replace(_re, '');
}


export function normalisedURLHostname(url: Url): string {
    const _hostname = new URL(url).hostname;
    const hostname = normaliseHostname(_hostname);
    return hostname;
}
